src/goto-instrument/contracts/contracts.cpp

```diff
@@ -430,10 +444,41 @@ void code_contractst::check_apply_loop_contracts(
     }
 
     goto_function.body.destructive_insert(
-      loop_body, add_pragma_disable_assigns_check(generated_code));
+      loop_head, add_pragma_disable_assigns_check(generated_code));
   }
 
-  // Generate: assert(invariant) just after the loop exits
+  // Insert all the generated code so far:
+  // Assume invariant, decl the variant temporaries,
+  // and initialize old variant values.
+  goto_function.body.destructive_insert(
```

Comment:
Why do we use three distinct destructive inserts instead of one ? it seems like the results would be exacly the same

Reply:
This is because a block of code in the middle (the havocing code) shouldn't be pragma-fied. The block above and below it should be (and their order matters), to disable assigns clause checking of the enclosing block on them. So right now the block is split into 3 blocks with distinct pragma annotations and we call insert thrice.
I added lots of comments now and this would hopefully be clear in the code now.